runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11658,12 +11659,106 @@ J9::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&result
+
+            //if offheap is not enabled, we will never need to load dataAddr. So we know for sure that we can calculate the final
+            //destination address right away and only pass in three children to setmemoryEvaluator() (dest, len, byteValue)
+            bool loadDataAddr = false;
+            bool separateDestAndOffset = false;
+
+         #if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+            if (TR::Compiler->om.isOffHeapAllocationEnabled() && comp->target().is64Bit())
+               {
+               if (dest->getSymbolReference() != NULL)
+                  {
+                  //check dstBaseAddrNode type at compile time
+                  int length;
+                  const char *objTypeSig = dest->getSymbolReference()->getTypeSignature(length);
+
+                  // There are four cases where we cannot be sure of the Object type at compile time:
+                  // 1.) The object's type signature is unavailable/unknown
+                  // 2.) The signature names java/lang/Object
+                  // 3.) The signature names some interface type, wich is a problem because interface types are
+                  //     not checked by the verifier
+                  // 4.) The signature belongs to a parameter of the calling method, which is a problem because
+                  //     parameters can share slots with variables of other, potentially incompatible types
+                  bool objTypeUnknown;
+                  TR_OpaqueClasBlock *objClass = NULL;
+
+                  if (objTypeSig == NULL)
+                     {
+                     objTypeUnknown = true;
+                     }
+                  else
+                     {
+                     objClass = fej9->getClassFromSignature(objTypeSig, length, dest->getSymbolReference()->getOwningMethod(comp));
+                     bool isParameter = dest->getSymbolReference()->getSymbol()->isParm();
+
+                     objTypeUnknown = objClass == NULL || objClass == comp->getObjectClassPointer() ||
+                                      TR::Compiler->cls.isInterfaceClass(comp, objClass) || isParameter;
+                     }
+
+                  //need to generate code to loadDataAddr in cases (2) and (3) (unless the object is known to be NULL)
+                  loadDataAddr = !dest->isNull() && (objTypeUnknown || objTypeSig[0] == '[');
+
+                  //pass in dest and destOffset separately in case (3) only
+                  separateDestAndOffset = loadDataAddr && objTypeUnknown;
+                  }
+               else
+                  {
+                  //if for some reason we cannot get the symbol reference (and thus, the type signature)
+                  //of dest, we'll need to take the conservative approach and pass dest and destOffset to
+                  //setmemoryEvaluator() separately
+                  separateDestAndOffset = true;
+                  loadDataAddr = true;
```

Comment:
later, it appeared setting loadDataAddr here is irrelevant anyway?

Reply:
This is only in the case where, for whatever reason, we can't get the symbol reference of the object (i.e.: `dest->getSymbolReference() == NULL`). In that case, since we can't get any information about the type of the object, we take the safe approach and pass `dest` and `destOffset` into `setmemoryEvaluator()` separately so that a runtime array check can be generated.